File: scripts/qgis/extract_styles_from_qgz.mjs
/**
 * Copia simbologia ORIGINAL des d'un projecte QGIS (.qgz)
 * cap als .qml de les capes a "retall 11-12".
 *
 * node scripts/qgis/extract_styles_from_qgz.mjs
 * node scripts/qgis/extract_styles_from_qgz.mjs "C:\CGIS\TFG8-qgis2\QGIS2WEB.qgz"
 */

import fs from "node:fs";
import path from "node:path";
import { execFileSync } from "node:child_process";

const RETALL_DIR = String.raw`C:\CGIS\TFG9-P11-12\retall 11-12`;
const WORK = String.raw`C:\CGIS\TFG9-P11-12\_qgz_extract_styles`;

// Projectes amb simbologia bona (per ordre de preferència)
const DEFAULT_SOURCES = [
  String.raw`C:\CGIS\TFG8-qgis2\QGIS2WEB.qgz`,
  String.raw`C:\CGIS\TFG9-RETALL-PARADES\QGIS2WEB.qgz`,
  String.raw`C:\CGIS\TFG8-qgis2 - copia\QGIS2WEB.qgz`,
  String.raw`C:\CGIS\TFG9-P11-12\QGIS2WEB.qgz`,
];

/** clau normalitzada → stem fitxer a retall 11-12 */
const MAP = [
  ["contactes discordants", "P11-12_Contactes Discordants i Esllavissaments"],
  ["esllavissaments", "P11-12_Contactes Discordants i Esllavissaments"],
  ["contactes generals", "P11-12_Contactes Generals i Intrusius"],
  ["contactes intrusius", "P11-12_Contactes Generals i Intrusius"],
  ["diposits quaternaris", "P11-12_Dipòsits Quaternaris i Materials Superficials"],
  ["materials superficials", "P11-12_Dipòsits Quaternaris i Materials Superficials"],
  ["geomorfologica", "P11-12_Geomorfològica de Vessants"],
  ["geomorfologia de vessants", "P11-12_Geomorfològica de Vessants"],
  ["litologia del socol", "P11-12 Litologia del Sòcol (Paleozoic i Mesozoic)"],
  ["paleozoic i mesozoic", "P11-12 Litologia del Sòcol (Paleozoic i Mesozoic)"],
  ["tipus i naturalesa", "P11-12_Tipus i Naturalesa Litològica"],
  ["naturalesa litologica", "P11-12_Tipus i Naturalesa Litològica"],
  ["toponimia", "P11-12_Toponimia i Noms geogràfics"],
  ["noms geografics", "P11-12_Toponimia i Noms geogràfics"],
  ["interes geologic", "P11-12_Zones considerades d'interesgeologic"],
  ["zones considerades", "P11-12_Zones considerades d'interesgeologic"],
  ["espaisintgeol", "P11-12_Zones considerades d'interesgeologic"],
  ["espais d interes", "P11-12_Zones considerades d'interesgeologic"],
  ["corbes de nivell", "P11-12_Carbes de nivell"],
  ["carbes de nivell", "P11-12_Carbes de nivell"],
  ["ruta sau", "P11-12_Ruta Sau-Taveret"],
  ["sau taveret", "P11-12_Ruta Sau-Taveret"],
  ["xarxa viaria", "P11-12_Xarxa viària"],
  ["xarxa vi", "P11-12_Xarxa viària"],
  ["construccions", "P11-12_Contruccions"],
  ["contruccions", "P11-12_Contruccions"],
  ["parades p", "P11-12_Parades-P"],
  ["parades-p", "P11-12_Parades-P"],
  ["afloraments", "P11-12_Afloraments"],
  ["falles", "P11-12_Falles"],
  ["noms", "P11-12_Noms"],
  ["cobertes del sol", "Cobertes del sol 11-12"],
  ["cobertes", "Cobertes del sol 11-12"],
  ["ortofoto", "satelit11-12"],
  ["satelit", "satelit11-12"],
];

function norm(s) {
  return String(s || "")
    .toLowerCase()
    .normalize("NFD")
    .replace(/\p{M}/gu, "")
    .replace(/['’·]/g, "")
    .replace(/[\W_]+/g, " ")
    .replace(/\s+/g, " ")
    .trim();
}

function pickSource() {
  const arg = process.argv[2];
  if (arg && fs.existsSync(arg)) return arg;
  for (const p of DEFAULT_SOURCES) {
    if (fs.existsSync(p)) return p;
  }
  throw new Error("No s'ha trobat cap .qgz font d'estils");
}

function extractZip(qgz) {
  fs.mkdirSync(WORK, { recursive: true });
  for (const f of fs.readdirSync(WORK)) {
    fs.rmSync(path.join(WORK, f), { recursive: true, force: true });
  }
  const zipPath = path.join(WORK, "proj.zip");
  fs.copyFileSync(qgz, zipPath);
  execFileSync(
    "powershell.exe",
    [
      "-NoProfile",
      "-Command",
      `Expand-Archive -LiteralPath '${zipPath.replace(/'/g, "''")}' -DestinationPath '${WORK.replace(/'/g, "''")}' -Force`,
    ],
    { stdio: "inherit" },
  );
  const qgs = fs.readdirSync(WORK).find((f) => f.endsWith(".qgs"));
  if (!qgs) throw new Error("No .qgs dins del qgz");
  return path.join(WORK, qgs);
}

function matchStem(layername, datasource) {
  const blob = norm(`${layername} ${datasource}`);
  // ignora el polígon de retall genèric
  if (blob === "retall" || blob.endsWith(" retall") && blob.length < 12) {
    return null;
  }
  for (const [key, stem] of MAP) {
    if (blob.includes(norm(key))) return stem;
  }
  return null;
}

function findRetallFile(stem) {
  const target = norm(stem);
  const files = fs
    .readdirSync(RETALL_DIR)
    .filter((f) => /\.(gpkg|shp|tif|tiff)$/i.test(f));
  for (const f of files) {
    if (norm(path.parse(f).name) === target) return path.join(RETALL_DIR, f);
  }
  for (const f of files) {
    const stemF = norm(path.parse(f).name);
    if (stemF.includes(target) || target.includes(stemF)) {
      return path.join(RETALL_DIR, f);
    }
  }
  return null;
}

function extractMapLayers(xml) {
  const layers = [];
  const re = /<maplayer\b[\s\S]*?<\/maplayer>/g;
  let m;
  while ((m = re.exec(xml))) {
    const block = m[0];
    const layername = (block.match(/<layername>([^<]*)<\/layername>/) || [])[1] || "";
    const datasource =
      (block.match(/<datasource>([^<]*)<\/datasource>/) || [])[1] || "";
    const rendererType =
      (block.match(/<renderer-v2[^>]*type="([^"]+)"/) || [])[1] || "";
    const categories = (block.match(/<category /g) || []).length;
    const hasRenderer =
      block.includes("<renderer-v2") || block.includes("<pipe>");
    layers.push({
      block,
      layername,
      datasource,
      hasRenderer,
      rendererType,
      categories,
      bytes: block.length,
    });
  }
  return layers;
}

function maplayerToQml(block) {
  let body = block
    .replace(/^<maplayer\b/, "<qgis")
    .replace(/<\/maplayer>$/, "</qgis>");
  if (!/styleCategories=/.test(body)) {
    body = body.replace("<qgis", '<qgis styleCategories="AllStyleCategories"');
  }
  return `<!DOCTYPE qgis PUBLIC 'http://mrcc.com/qgis.dtd' 'SYSTEM'>\n${body}\n`;
}

function scoreLayer(layer) {
  // preferim categorized amb moltes categories / blocs més grans
  let s = layer.bytes;
  if (layer.rendererType === "categorizedSymbol") s += 100000 + layer.categories * 1000;
  if (layer.rendererType === "graduatedSymbol") s += 50000;
  if (layer.rendererType === "RuleRenderer") s += 80000;
  return s;
}

function main() {
  const qgz = pickSource();
  console.log("Font d'estils:", qgz);
  console.log("Destí retalls:", RETALL_DIR);

  const qgsPath = extractZip(qgz);
  const xml = fs.readFileSync(qgsPath, "utf8");
  const layers = extractMapLayers(xml);
  console.log("Capes al projecte font:", layers.length);

  // Per cada stem destí, queda't la millor capa font
  const best = new Map(); // stem -> layer
  for (const layer of layers) {
    if (!layer.hasRenderer) continue;
    const stem = matchStem(layer.layername, layer.datasource);
    if (!stem) continue;
    const prev = best.get(stem);
    if (!prev || scoreLayer(layer) > scoreLayer(prev)) {
      best.set(stem, layer);
    }
  }

  let written = 0;
  for (const [stem, layer] of best.entries()) {
    const retallFile = findRetallFile(stem);
    if (!retallFile) {
      console.log(`· Sense fitxer retall: [${layer.layername}] → ${stem}`);
      continue;
    }
    const qmlPath = retallFile.replace(/\.[^.]+$/i, ".qml");
    if (fs.existsSync(qmlPath)) {
      fs.copyFileSync(qmlPath, `${qmlPath}.bak`);
    }
    fs.writeFileSync(qmlPath, maplayerToQml(layer.block), "utf8");
    written++;
    console.log(
      `✓ [${layer.layername}] (${layer.rendererType || "raster"}, cat=${layer.categories}) → ${path.basename(qmlPath)}`,
    );
  }

  // Copia també Tipus v2 si existeix
  const tipusQml = path.join(RETALL_DIR, "P11-12_Tipus i Naturalesa Litològica.qml");
  const tipusV2 = path.join(RETALL_DIR, "P11-12 Tipus i naturalesa litologica v2.qml");
  if (fs.existsSync(tipusQml) && fs.existsSync(path.join(RETALL_DIR, "P11-12 Tipus i naturalesa litologica v2.gpkg"))) {
    fs.copyFileSync(tipusQml, tipusV2);
    console.log("✓ Copiat estil Tipus → v2");
  }

  console.log(`\nFet. QML escrits: ${written}`);
  console.log("Ara a QGIS executa:");
  console.log(
    '  exec(open(r"C:/kotei-tfg/scripts/qgis/aplicar_qml_retalls.py", encoding="utf-8").read())',
  );
}

main();

File: scripts/qgis/extract_styles_from_qgz.py
"""
Extreu estils del projecte QGIS2WEB.qgz i els desa com a .qml
al costat de cada capa retallada a 'retall 11-12'.

No cal QGIS en execució: llegeix l'XML del .qgs.
"""

from __future__ import annotations

import re
import shutil
import zipfile
from pathlib import Path
from xml.etree import ElementTree as ET

QGZ = Path(r"C:\CGIS\TFG9-P11-12\QGIS2WEB.qgz")
RETALL_DIR = Path(r"C:\CGIS\TFG9-P11-12\retall 11-12")
WORK = Path(r"C:\CGIS\TFG9-P11-12\_qgz_extract")

# Nom capa al projecte (o fragment) → stem del fitxer retall
MAP = {
    "afloraments": "P11-12_Afloraments",
    "parades": "P11-12_Afloraments",  # a vegades el layername intern
    "carbes de nivell": "P11-12_Carbes de nivell",
    "corbes de nivell": "P11-12_Carbes de nivell",
    "contactes discordants": "P11-12_Contactes Discordants i Esllavissaments",
    "contactes generals": "P11-12_Contactes Generals i Intrusius",
    "contruccions": "P11-12_Contruccions",
    "construccions": "P11-12_Contruccions",
    "dipòsits quaternaris": "P11-12_Dipòsits Quaternaris i Materials Superficials",
    "diposits quaternaris": "P11-12_Dipòsits Quaternaris i Materials Superficials",
    "falles": "P11-12_Falles",
    "geomorfològica": "P11-12_Geomorfològica de Vessants",
    "geomorfologica": "P11-12_Geomorfològica de Vessants",
    "litologia del sòcol": "P11-12 Litologia del Sòcol (Paleozoic i Mesozoic)",
    "litologia del socol": "P11-12 Litologia del Sòcol (Paleozoic i Mesozoic)",
    "noms": "P11-12_Noms",
    "parades-p": "P11-12_Parades-P",
    "parades p": "P11-12_Parades-P",
    "ruta sau": "P11-12_Ruta Sau-Taveret",
    "cami-sau": "P11-12_Ruta Sau-Taveret",
    "tipus i naturalesa": "P11-12_Tipus i Naturalesa Litològica",
    "toponimia": "P11-12_Toponimia i Noms geogràfics",
    "toponímia": "P11-12_Toponimia i Noms geogràfics",
    "xarxa vi": "P11-12_Xarxa viària",
    "zones considerades": "P11-12_Zones considerades d'interesgeologic",
    "interes geologic": "P11-12_Zones considerades d'interesgeologic",
    "interès geològic": "P11-12_Zones considerades d'interesgeologic",
    "espaisintgeol": "P11-12_Zones considerades d'interesgeologic",
    "cobertes": "Cobertes del sol 11-12",
    "satelit": "satelit11-12",
    "ortofoto": "satelit11-12",
}


def norm(s: str) -> str:
    s = (s or "").strip().lower()
    repl = {
        "à": "a",
        "á": "a",
        "è": "e",
        "é": "e",
        "í": "i",
        "ï": "i",
        "ò": "o",
        "ó": "o",
        "ú": "u",
        "ü": "u",
        "ç": "c",
        "·": "",
        "'": "",
        "’": "",
    }
    for a, b in repl.items():
        s = s.replace(a, b)
    s = re.sub(r"[\W_]+", " ", s)
    return re.sub(r"\s+", " ", s).strip()


def extract_qgs() -> Path:
    WORK.mkdir(parents=True, exist_ok=True)
    with zipfile.ZipFile(QGZ, "r") as zf:
        zf.extractall(WORK)
    qgs_files = list(WORK.glob("*.qgs"))
    if not qgs_files:
        raise FileNotFoundError("No s'ha trobat .qgs dins del .qgz")
    return qgs_files[0]


def match_retall_stem(layername: str, datasource: str) -> str | None:
    blob = norm(layername + " " + datasource)
    # primer matches més específics (claus més llargues)
    for key in sorted(MAP.keys(), key=len, reverse=True):
        if norm(key) in blob:
            return MAP[key]
    return None


def maplayer_to_qml_xml(maplayer: ET.Element) -> str:
    """Converteix un <maplayer> del projecte en un document .qml vàlid."""
    # El .qml és bàsicament el maplayer amb root <qgis>
    attrs = dict(maplayer.attrib)
    # Assegura categories d'estil
    attrs.setdefault("styleCategories", "AllStyleCategories")

    # Clona profundament
    root = ET.Element("qgis", attrs)
    for child in list(maplayer):
        root.append(child)

    # Declaració
    xml = ET.tostring(root, encoding="unicode")
    return "<!DOCTYPE qgis PUBLIC 'http://mrcc.com/qgis.dtd' 'SYSTEM'>\n" + xml + "\n"


def find_retall_file(stem: str) -> Path | None:
    """Troba .gpkg/.shp/.tif amb aquest stem (tolerant a accents/espais)."""
    target = norm(stem)
    for path in RETALL_DIR.iterdir():
        if not path.is_file():
            continue
        if path.suffix.lower() not in {".gpkg", ".shp", ".tif", ".tiff", ".png", ".jpg"}:
            continue
        if norm(path.stem) == target:
            return path
    # partial
    for path in RETALL_DIR.iterdir():
        if path.suffix.lower() not in {".gpkg", ".shp", ".tif", ".tiff"}:
            continue
        if target in norm(path.stem) or norm(path.stem) in target:
            return path
    return None


def main() -> None:
    qgs_path = extract_qgs()
    print("Projecte:", qgs_path)

    # Parse tolerant a namespaces
    text = qgs_path.read_text(encoding="utf-8", errors="replace")
    # ElementTree amb maplayer
    root = ET.fromstring(text)

    maplayers = root.findall(".//maplayer")
    print(f"Capes al projecte: {len(maplayers)}")

    written = 0
    skipped = 0
    used_stems: set[str] = set()

    for ml in maplayers:
        layername_el = ml.find("layername")
        ds_el = ml.find("datasource")
        layername = layername_el.text if layername_el is not None else ""
        datasource = ds_el.text if ds_el is not None else ""

        stem = match_retall_stem(layername or "", datasource or "")
        if not stem:
            skipped += 1
            continue

        # Evita sobreescriure el mateix retall amb una capa menys específica
        if stem in used_stems:
            # Preferim la primera coincidència (ja ordenada per especificitat de clau)
            continue

        retall_file = find_retall_file(stem)
        if not retall_file:
            print(f"· Sense fitxer retall per '{layername}' → {stem}")
            continue

        # Comprova que hi ha renderer
        if ml.find("renderer-v2") is None and ml.find("pipe") is None:
            # ràster usa pipe; vector renderer-v2
            print(f"· Sense renderer: {layername}")
            continue

        qml_path = retall_file.with_suffix(".qml")
        # Backup de l'qml anterior si existeix
        if qml_path.exists():
            bak = qml_path.with_suffix(".qml.bak")
            shutil.copy2(qml_path, bak)

        qml_xml = maplayer_to_qml_xml(ml)
        qml_path.write_text(qml_xml, encoding="utf-8")
        used_stems.add(stem)
        written += 1
        print(f"✓ [{layername}] → {qml_path.name}")

    print(f"\nFet. QML escrits: {written}. Capes projecte sense match: {skipped}")
    print(
        "A QGIS: treu i torna a afegir les capes de 'retall 11-12', "
        "o clic dret → Estilos → Cargar estilo → el .qml del costat."
    )


if __name__ == "__main__":
    main()
